### PR TITLE
[Feature] Restricted Assets 

### PR DESCRIPTION
**Pull request by blondfrogs**

### CODE REVIEW

File: src/assets/assets.cpp [hunk 1]
```diff
@@ -4680,11 +4735,15 @@ bool CAssetsCache::CheckForAddressQualifier(const std::string &qualifier_name, c
     }
 
     if (prestricteddb) {
-        if (prestricteddb->ReadAddressQualifier(address, qualifier_name)) {
-            if (passetsQualifierCache)
-                passetsQualifierCache->Put(cachedQualifierAddress.GetHash().GetHex(), 1);
+        if (prestricteddb->CheckForAddressRootQualifier(address, qualifier_name)){
             return true;
         }
+//                } else {
```
Comment (cfox) [inline src/assets/assets.cpp]: Remove comment block?

File: src/rpc/assets.cpp [hunk 2]
```diff
@@ -1385,21 +1721,951 @@ UniValue getcacheinfo(const JSONRPCRequest& request)
...
+{
+    if (request.fHelp || !AreRestrictedAssetsDeployed() || request.params.size() < 2 || request.params.size() > 3)
+        throw std::runtime_error(
+                "freezeaddress asset_name address (change_address)\n"
+                + RestrictedActivationWarning() +
+                "\nFreeze an address from transferring a restricted asset\n"
+
+                "\nArguments:\n"
+                "1. \"asset_name\"       (string, required) the name of the restricted asset you want to freeze\n"
+                "2. \"address\"          (string), required) the address that will be frozen\n"
```
Comment (TronBlack) [inline src/rpc/assets.cpp]: Strange formatting with parens (next line too)

File: src/rpc/assets.cpp [hunk 3]
```diff
@@ -1385,21 +1721,951 @@ UniValue getcacheinfo(const JSONRPCRequest& request)
...
+{
+    if (request.fHelp || !AreRestrictedAssetsDeployed() || request.params.size() < 2 || request.params.size() > 3)
+        throw std::runtime_error(
+                "unfreezeaddress asset_name address (change_address)\n"
+                + RestrictedActivationWarning() +
+                "\nUnfreeze an address from transferring a restricted asset\n"
+
+                "\nArguments:\n"
+                "1. \"asset_name\"       (string, required) the name of the restricted asset you want to unfreeze\n"
+                "2. \"address\"          (string), required) the address that will be unfrozen\n"
```
Comment (TronBlack) [inline src/rpc/assets.cpp]: Same strange formatting with parens

File: src/rpc/assets.cpp [hunk 4]
```diff
@@ -1385,21 +1721,951 @@ UniValue getcacheinfo(const JSONRPCRequest& request)
...
+UniValue checkglobalrestriction(const JSONRPCRequest& request)
+{
+    if (request.fHelp || !AreRestrictedAssetsDeployed() || request.params.size() != 1)
+        throw std::runtime_error(
+                "checkglobalrestriction restricted_name\n"
+                + RestrictedActivationWarning() +
+                "\nChecks to see if a restricted asset is globally frozen\n"
+
+                "\nArguments:\n"
+                "1. \"restricted_name\"   (string), required) the restricted asset to search\n"
```
Comment (TronBlack) [inline src/rpc/assets.cpp]: parens

File: src/test/assets/asset_tx_tests.cpp [hunk 5]
```diff
@@ -115,7 +118,7 @@ BOOST_FIXTURE_TEST_SUITE(asset_tx_tests, BasicTestingSetup)
         // The outputs are assigning a destination to only 100 Assets
         // This should fail because 900 Assets aren't being assigned a destination (Trying to burn 900 Assets)
         std::vector<std::pair<std::string, uint256>> vReissueAssets;
-        BOOST_CHECK_MESSAGE(!Consensus::CheckTxAssets(tx, state, coins, vReissueAssets, true), "CheckTxAssets should of failed");
+        BOOST_CHECK_MESSAGE(!Consensus::CheckTxAssets(tx, state, coins, nullptr, false, vReissueAssets, true), "CheckTxAssets should of failed");
```
Comment (TronBlack) [inline src/test/assets/asset_tx_tests.cpp]: should've or "should have" (super nitpicky)

File: src/assets/assets.cpp [hunk 6]
```diff
@@ -2098,6 +2966,8 @@ size_t CAssetsCache::GetCacheSize() const
     size += (80 + 40 + 32 + 32 + sizeof(int)) * setNewReissueToAdd.size(); // CReissueAsset, Address, COutPoint, Block hash, int
     size += (80 + 40 + 32 + 32 + sizeof(int)) * setNewReissueToRemove.size(); // CReissueAsset, Address, COutPoint, Block hash, int
 
+    // TODO add the qualfier, and restricted sets into this calculation
```
[the remark this answers is not in the document]
Comment (blondfrogs) [reply src/assets/assets.cpp]: Yep.

File: src/assets/assetdb.h [hunk 7]
```diff
@@ -23,6 +25,9 @@ struct CBlockAssetUndo
     bool fChangedUnits;
     std::string strIPFS;
     int nUnits;
```
[the remark this answers is not in the document]
Comment (blondfrogs) [reply src/assets/assetdb.h]: Changed

**Pull request merged**